src/libstore/outputs-spec.hh

```diff
@@ -1,32 +1,95 @@
 #pragma once
 
+#include <optional>
+#include <set>
 #include <variant>
 
-#include "util.hh"
-
 #include "nlohmann/json_fwd.hpp"
 
 namespace nix {
 
-typedef std::set<std::string> OutputNames;
+struct OutputNames : std::set<std::string> {
```

Comment:
the STL container classes are *not* meant to be inherited from. Please don't.

Reply:
Is it a problem when it's a "newtype" like class --- exactly one superclass/field and so hopefully the same representation?

Reply:
Hm... I see that you try to make it harder to create an empty set. 
Unfortunately I wouldn't deem this approach to be very effective since you can still write `auto myEmptyInstance = OutputNames({});` and it does not even look wrong, because the typename `OutputNames` does not convey that it's meant to never be empty. For this reason, one still has to put the same effort into the review to find out if we really never construct non-empty sets.

From a design perspective, this is not a case for inheritance.
I suggest to either drop this subclass which is from a design perspective more harm than good, or implement a class `template <typename T> class NonEmptySet` that *has* an `std::set` and that really implements the semantics of a non-empty data structure.

Reply:
For what it's worth, there was one "builder pattern" type situation where I did need to start with the empty set as an identity for merging proper ones in. So I dont really mind that there is an escape hatch. Of course yes the best thing to do would be to use a separate type for this one instance and make it ironclad.